### DISCUSSION TITLE
User.paid returns true but subscription status is cancelled

### DISCUSSION THREAD

**Discussion by takkbbl** [Glench/ExtPay#146] [Q&A] [answered]
I was reading the docs and trying to find out why this happens but I couldn't understand it. I have a if statement that tests whether user.paid is true or false. In test mode, I changed the subscription status of my user to cancelled, but user.paid still returns true.

Any ideas?

![Screenshot 2023-06-23 at 22 48 13](https://github.com/Glench/ExtPay/assets/63558626/93bd217a-f39f-4250-899e-0716bd530037)

**Comment by Glench** [comment]
If you'd like support from us, would you please describe the precise steps to reproduce the error you're seeing or else paste the snippets of code you're using in your screenshots?

**Reply by takkbbl** [reply]
Below is a simplification of the code I use:

`extpay.getUser().then(user => { if (user.paid) { user_paid = user.paidAt; console.log("user paid:"); console.log(user.paid); // User paid } else if (user.trialStartedAt && (now - user.trialStartedAt) < sevenDays) { // User is on a free trial }  else if (user.trialStartedAt && (now - user.trialStartedAt) >= sevenDays) { // Free trial ended } else{ //new user } }).catch(err => { //error })`

There is not much I can add to this. I just changed the subscription status to canceled, ran the code and it runs into user.paid = true.

**Reply by Glench** [reply]
Is that code in the background file or a popup or a content script?

**Reply by askalia** [reply]
I got same in dev. 
Occurs in background
I guess this is because of : even though you force unsubscribe state, the current month still is considered paid so it will be marked as false only as of the next month's day. (If I'm right) 
In prod that's ok, in dev, unsubscribe should have forced-state be effective immediately.

**Reply by takkbbl** [reply]
For me in occurs in a content script.

**Reply by Glench** [reply, marked as answer]
Hm, I'm not able to reproduce what you're seeing. I tested and it worked fine. Would you be able to make a minimal repository and instructions that I can run precisely to reproduce the issue?

**Reply by connallykendrick** [reply]
I'm having this issue currently. I'd like to force a "Canceled" status, but instead it's returning true as if I'm still within the month of paid benefits.

It'd be great to have a "Canceled - Pending" & the "Canceled" option to test all.

i.e. - else if (user.subscriptionCancelAt && user.subscriptionCancelAt < new Date()) {
        console.log("Your subscription will cancel soon.")
    } else if (user.subscriptionStatus === 'canceled') {
        console.log("We hope you enjoyed your subscription!")

**Reply by Glench** [reply]
@connallykendrick We do have this! You can just click the normal "cancel" button that your users would see. You can also manually set canceled status in dev mode.

**Reply by connallykendrick** [reply]
> @connallykendrick We do have this! You can just click the normal "cancel" button that your users would see. You can also manually set canceled status in dev mode.
> 

Wow. Completely overlooked that. Lol Thank you! I'll test it out

**Reply by Glench** [reply]
No problem. And yeah, I think it's an easy mistake to make and I believe I already made it even clearer in the next unreleased version that I've been testing.

**Reply by connallykendrick** [reply]
Awesome! Excited to see the new.
Since, I've been having trouble getting it to trigger what I need. Everything else seems to work great. I must be overlooking something - I'll test more and get back to you.

**Reply by Swansocl1** [reply]
![image](https://github.com/Glench/ExtPay/assets/120823142/374ac125-136d-413e-8aa8-df3890949ad9)

![image](https://github.com/Glench/ExtPay/assets/120823142/5b2a92ee-82c0-4e90-bc34-375846982625)

Hey Glen, I'm having a similar issue. In dev mode I forced my status to cancelled, and it notes that paid features were disabled but in the console log paid status is still returning true. So I'm unable to test the functionality with a non-paid/free trial status because with any other email I use I can't login without paying or using the free trial. Any assistance would be appreciated.

**Reply by Glench** [reply]
hey @Swansocl1 can you make a minimal repository demonstrating the issue for me with instructions for how to reproduce it? That is the best way to track down what's happening and help you out.

**Reply by ba32107** [reply]
I have the same issue as the others. I'm testing in dev, and no matter what subscription status I set in dev mode, the `user.paid` is always true and there is never a `subscriptionStatus` or `subscriptionCancelAt` property on the object.

I'll look into creating a small repro repo.

In general, I find the dev workflow a bit hard and it's not very clear to me what's happening. For example, I just logged in with my email and it seems like ExtensionPay somehow "knows" that I already done this testing workflow before, because my status was paid (like I said above, it's always paid). Is my email (the one that I used for registering on extensionpay.com) treated as special in any way? I don't really understand what's happening.

What I need is the ability to reset everything. I want to clear the free trial, clear any payment info, on your server side, everything, so I can test my workflows from scratch (if I was a user who sees my extension for the first time). And I want to do this multiple times until I get my extension's code right. I also don't want to use another email address, because I don't want to create multiple emails just for this.

So the best would be a dashboard on the website that provides visibility on the testing process. Alternatively, a bit better docs that explain what happens on the server side would also be helpful.

**Reply by ba32107** [reply]
Okay I managed to reproduce this. I'm pretty sure it's related to the login functionality. It's hard to tell if this is an actual bug, or just a byproduct of the special handling of my developer email account. This just reiterates what I wrote in my previous comment: things like this makes the developer workflow quite tedious and I can never be sure, is what I'm experiencing during development the same thing the end user is gonna see? What exactly works differently with my developer account?

Note, this is all meant as constructive feedback and not pointless criticism - hopefully you don't take it as such! 

Anyway here is the code for the repro:
https://github.com/ba32107/ExtPay/tree/subscription_Status_repro/sample-extension-mv3
Here's a PR so you can see the diff more easily:
https://github.com/ba32107/ExtPay/pull/1/files

Repro steps:
1. open popup
2. click `Manage Payments`
3. sign up with your dev email
4. close payments page
5. click `Refresh`, observe everything is as expected
6. `Manage payments`, change subscription status to something else, submit, `Refresh`, observe everything works as expected
7. Now log out
8. `Manage payments` and log in (don't subscribe) with your dev email
9. click `Refresh`, observe that `paid` is true and there is no `subscriptionStatus` property
10. `Manage payments` again, change subscription status to something else (e.g. `past_due`), submit
11. click `Refresh`, and again observe that `paid` is true and there is no `subscriptionStatus` property

**Reply by Glench** [reply]
Hey @ba32107 we don't have a log out function! So it's no wonder you're seeing something incorrect. If you're able to reproduce this in the new version of ExtPay (3.1.1) let me know.

**Reply by ba32107** [reply]
The logout is something I implemented in my extension (and this small repro as well):
https://github.com/ba32107/ExtPay/blob/subscription_Status_repro/sample-extension-mv3/popup.js#L18-L23

It's been a while since I made this comment and I haven't had much time to work on my extension lately, but if memory serves right, there was inconsistent behavior here. "logout" is something I implemented to simulate a user installing the extension for the first time.

When I get a chance to work on this again I'll definitely give it another go with the new version!

**Reply by Glench** [reply]
OK thanks for responding! Yeah, just so you know we don't officially support unsanctioned modifications like this. I'll do my best to understand and help, but if you're doing something not intended then you should know that.
